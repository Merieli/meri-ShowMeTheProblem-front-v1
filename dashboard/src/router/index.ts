import { createRouter, createWebHistory, RouteRecordRaw } from 'vue-router';

const Home = () => import('../views/Home/PageHome.vue');
const Feedbacks = () => import('../views/Feedbacks/PageFeedbacks.vue');
const Credentials = () => import('../views/Credentials/PageCredentials.vue');

export const routes: Array<RouteRecordRaw> = [
    {
        path: '/',
        name: 'home',
        component: Home,
    },
    {
        path: '/feedbacks',
        name: 'Feedbacks',
        component: Feedbacks,
        meta: {
            hasAuth: true,
        },
    },
    {
        path: '/Credentials',
        name: 'Credentials',
        component: Credentials,
        meta: {
            hasAuth: true,
        },
    },
    {
        path: '/:pathMatch(.*)*',
        redirect: { name: 'Home' },
    },
];

const router = createRouter({
    history: createWebHistory('/'),
    routes,
});

export default router;
