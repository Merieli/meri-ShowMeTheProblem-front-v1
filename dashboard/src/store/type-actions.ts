export enum Actions {
    LOGIN_USER = 'LOGIN_USER',
    REGISTER_USER = 'REGISTER_USER',
    GET_USER = 'GET_USER',

    NEW_FEEDBACK = 'NEW_FEEDBACK',
    GET_ALL_FEEDBACKS = 'GET_ALL_FEEDBACKS',
    GET_FEEDBACKS = 'GET_FEEDBACKS',
    GET_INDEX_FEEDBACK = 'GET_INDEX_FEEDBACK',
    CHANGE_ACTIVE_FEEDBACK = 'CHANGE_ACTIVE_FEEDBACK',
}
