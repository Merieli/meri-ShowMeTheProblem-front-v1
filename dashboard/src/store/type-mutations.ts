export enum Mutations {
    NOTIFY = 'NOTIFY',

    ADD_USER = 'ADD_USER',
    LOGIN_USER = 'LOGIN_USER',
    USER_LOGGED = 'USER_LOGGED',

    ADD_FILTERS = 'ADD_FILTERS',
    ADD_CONFIGURED_FILTERS = 'ADD_CONFIGURED_FILTERS',
    ADD_FEEDBACKS = 'ADD_FEEDBACKS',

    TOOGLE_LOADING = 'TOOGLE_LOADING',
}
