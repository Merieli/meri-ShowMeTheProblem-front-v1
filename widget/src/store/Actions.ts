export enum Actions {
    SAVE_FEEDBACK = 'SAVE_FEEDBACK',
}
