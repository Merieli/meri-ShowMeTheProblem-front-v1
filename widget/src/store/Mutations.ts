export enum Mutations {
    SET_CURRENT_COMPONENT = 'SET_CURRENT_COMPONENT',
    SET_MESSAGE = 'SET_MESSAGE',
    SET_FEEDBACK_TYPE = 'SET_FEEDBACK_TYPE',
    SET_CURRENT_PAGE = 'SET_CURRENT_PAGE',
    SET_APIKEY = 'SET_APIKEY',
    SET_FINGERPRINT = 'SET_FINGERPRINT',
    RESET_STORE = 'RESET_STORE',
}
