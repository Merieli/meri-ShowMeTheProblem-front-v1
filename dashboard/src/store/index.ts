import useNotifier from '../hooks/notifier'
import router from '../router/index'
import { Actions } from './type-actions'
import { Mutations } from './type-mutations'
import callApiClient from '@/http-client/index'
import { TypeOfNotification } from '@/interfaces/INotification'
import { INotification } from '@/interfaces/INotification'
import IUser from '@/interfaces/IUser'
import IUserLogged from '@/interfaces/IUserLogged'
import { createStore } from 'vuex'

export interface EstadoStore {
    users: IUser[]
    notifications: INotification[]
    isLogged: boolean
    userLogged: IUserLogged
}

export const store = createStore<EstadoStore>({
    state: {
        users: [],
        notifications: [],
        isLogged: false,
        userLogged: {
            name: '',
            token: '',
            apiKey: '',
        },
    },
    getters: {},
    mutations: {
        [Mutations.NOTIFY](state, newNotification: INotification) {
            newNotification.id = new Date().getTime()
            state.notifications.push(newNotification)
            setTimeout(() => {
                state.notifications = state.notifications.filter((notif) => notif.id != newNotification.id)
            }, 5000)
        },
        [Mutations.ADD_USER](state, user: IUser) {
            state.users.push(user)
        },
        [Mutations.LOGIN_USER](state, token) {
            state.userLogged.token = token
            state.isLogged = true
        },
        [Mutations.USER_LOGGED](state, data) {
            state.userLogged.name = data.name
            state.userLogged.apiKey = data.apiKey
        },
    },
    actions: {
        /**
         * @name REGISTER_USER
         * @descripton efetua o registro do usuário e o adiciona a store
         * @param {IUser} user usuário com name, email e password
         */
        async [Actions.REGISTER_USER]({ commit }, user: IUser) {
            try {
                const response = await callApiClient.user.register(user.name, user.email, user.password)
                commit(Mutations.ADD_USER, response)
            } catch (e) {
                throw new Error(`Não foi possível efetuar cadastro. Confira os dados informados`)
            }
        },

        /**
         * @name LOGIN_USER
         * @description Efetua login do usuário, salva o token na store,
         * redireciona para a página de 'feedbacks' e aciona a action
         * para pegar os dados do usuário logado.
         * @param {IUser} user usuário com name, email e password
         */
        async [Actions.LOGIN_USER]({ commit, dispatch, state }, user: IUser) {
            try {
                const response = await callApiClient.user.login(user.email, user.password)
                commit(Mutations.LOGIN_USER, response.token)
                router.push('/feedbacks')
                dispatch(Actions.GET_USER, state.userLogged.token)
            } catch (e) {
                const { notify } = useNotifier()
                notify(TypeOfNotification.FALHA, 'Não foi possível efetuar login', 'Confira os dados informados.')
                throw new Error(`Não foi possível efetuar login. Confira os dados informados`)
            }
        },

        /**
         * @name GET_USER
         * @descripton Captura os dados do usuários e salva na store
         * @param token chave de acesso do usuário
         */
        async [Actions.GET_USER]({ commit }, token) {
            try {
                const response = await callApiClient.user.show(token)
                commit(Mutations.USER_LOGGED, response)
            } catch (e) {
                throw new Error(`Não foi possível capturar os dados do usuário.`)
            }
        },
    },
    modules: {},
})
