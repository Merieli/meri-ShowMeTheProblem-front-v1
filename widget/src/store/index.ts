import httpClient from '@/http';
import { GettersStore, MutationsStore, StateStore } from '@/interfaces/';
import { ActionsStore } from '@/interfaces/ActionsStore';
import { InjectionKey } from 'vue';
import { createStore, Store, useStore as baseUseStore } from 'vuex';
import { Actions } from './Actions';
import { Mutations } from './Mutations';

export const state: StateStore = {
    currentComponent: 'SelectFeedbackType',
    message: '',
    feedbackType: '',
    fingerprint: '',
    apiKey: '',
    currentPage: '',
};

export const mutations: MutationsStore = {
    [Mutations.SET_CURRENT_COMPONENT](state, componentName): void {
        state.currentComponent = componentName;
    },
    [Mutations.SET_MESSAGE](state, message): void {
        state.message = message;
    },
    [Mutations.SET_FEEDBACK_TYPE](state, type): void {
        state.feedbackType = type;
    },
    [Mutations.SET_CURRENT_PAGE](state, page): void {
        state.currentPage = page;
    },
    [Mutations.SET_APIKEY](state, apiKey): void {
        state.apiKey = apiKey;
    },
    [Mutations.SET_FINGERPRINT](state, fingerprint): void {
        state.fingerprint = fingerprint;
    },
    [Mutations.RESET_STORE](state): void {
        state.currentComponent = 'SelectFeedbackType';
        state.message = '';
        state.feedbackType = '';
        state.currentPage = '';
        state.apiKey = '';
        state.fingerprint = '';
    },
};

export const getters: GettersStore = {
    currentComponent(state): string {
        return state.currentComponent;
    },
    feedbackType(state): string {
        return state.feedbackType;
    },
};

export const actions: ActionsStore = {
    async [Actions.SAVE_FEEDBACK]({ state, commit }, feedback): Promise<void> {
        try {
            commit(Mutations.SET_MESSAGE, feedback);
            await httpClient.feedbacks.create({
                type: state.feedbackType,
                text: state.message,
                page: state.currentPage,
                apiKey: state.apiKey,
                device: window.navigator.userAgent,
                fingerprint: state.fingerprint,
            });
            commit(Mutations.SET_CURRENT_COMPONENT, 'SuccessSave');
        } catch (error) {
            commit(Mutations.SET_CURRENT_COMPONENT, 'ErrorSave');
            throw new Error('Não foi possível salvar o feedback!');
        }
    },
};

export const store = createStore({
    state,
    getters,
    mutations,
    actions,
    modules: {},
});

export const key: InjectionKey<Store<StateStore>> = Symbol();

export function useStore(): Store<StateStore> {
    return baseUseStore(key);
}
