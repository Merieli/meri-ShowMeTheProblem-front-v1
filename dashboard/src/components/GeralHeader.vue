<template>
    <div class="bg-brand-main w-full justify-center flex">
        <header class="geral-header justify-between items-center flex">
            <h1 class="geral-header__logo font-black"><a href="#" @click="changePage('/')">ShowMeTheProblem</a></h1>
            <nav class="geral-header__navigation" v-if="!logged">
                <button id="button-create" class="geral-header__link font-bold" data-create @click="toggleModalCreate">
                    Crie uma conta
                </button>
                <BaseButton id="button-login" @click="toggleModalLogin" data-login> Entrar </BaseButton>
            </nav>
            <nav class="geral-header__navigation" v-else>
                <button class="geral-header__link font-bold" @click="changePage('/credentials')">Credenciais</button>
                <button class="geral-header__link font-bold" @click="changePage('/feedbacks')">Feedbacks</button>
                <BaseButton id="logout-button" @click="loggout()">{{ nameUser ? nameUser : '...' }} (sair)</BaseButton>
            </nav>
        </header>
        <PartModal type-modal="create" :open="openCreateModal" @close="toggleModalCreate" />
        <PartModal type-modal="login" :open="openLoginModal" @close="toggleModalLogin" />
    </div>
</template>

<script lang="ts">
/**
 * Cabeçalho para todas as páginas da aplicação.
 * @displayName Geral Header
 */
import router from '../router/index'
import { useStore } from '../store'
import { Actions } from '../store/type-actions'
import BaseButton from './BaseButton/index.vue'
import PartModal from './PartModal.vue'
import { computed, defineComponent, onMounted, ref } from 'vue'

export default defineComponent({
    name: 'GeralHeader',
    components: {
        BaseButton,
        PartModal,
    },
    setup() {
        const store = useStore()
        const openCreateModal = ref(false)
        const openLoginModal = ref(false)
        const logged = computed(() => store.state.isLogged)

        /**
         * Method Alterna o modal de criação de conta para ser oculto ou visivel
         */
        const toggleModalCreate = () => {
            openCreateModal.value = !openCreateModal.value
        }

        /**
         * Alterna o modal de login para ser oculto ou visivel
         */
        const toggleModalLogin = () => {
            openLoginModal.value = !openLoginModal.value
        }

        /**
         * Alterna a página atual
         * @param {string} rota caminho da página que será aberta ao clicar nos
         * botões do menu.
         */
        const changePage = (rota: string): void => {
            router.push(rota)
        }

        const loggout = (): void => {
            changePage('/')
            store.dispatch(Actions.LOGGOUT_USER)
        }

        /**
         * Captura os dados do usuário logado
         *
         */
        onMounted(() => {
            if (window.localStorage.getItem('token') && logged.value == false) {
                store.dispatch(Actions.GET_USER)
            }
        })

        return {
            openCreateModal,
            toggleModalCreate,
            openLoginModal,
            toggleModalLogin,
            changePage,
            loggout,
            logged,
            nameUser: computed(() => store.state.userLogged.name),
        }
    },
})
</script>

<style lang="scss" scoped>
.geral-header {
    height: 129px;
    padding: 1rem;
    width: 100%;
    max-width: 1080px;

    &__link {
        color: theme('colors.brand.white');
        padding-right: 2rem;
    }

    &__logo {
        color: theme('colors.brand.white');
    }
}
</style>
<docs lang="md">
## Examples

```jsx
<GeralHeader />
```
</docs>
