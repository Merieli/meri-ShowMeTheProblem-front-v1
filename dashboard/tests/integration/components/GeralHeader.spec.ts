import GeralHeader from '@/components/GeralHeader.vue'
import { routes } from '@/router'
import { key } from '@/store'
import { mount, VueWrapper } from '@vue/test-utils'
import { nextTick } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import { createStore } from 'vuex'

describe('GeralHeader', () => {
    const router = createRouter({
        history: createWebHistory(),
        routes,
    })

    const mockStore = createStore({
        state: {
            isLogged: false,
            userLogged: {
                name: '',
            },
        },
    })

    let wrapper: VueWrapper

    beforeEach(async () => {
        router.push('./')
        await router.isReady()

        wrapper = mount(GeralHeader, {
            global: {
                plugins: [[mockStore, key], router],
                stubs: {
                    PartModal: {
                        template: '<div>Modal de login e registro</div>',
                    },
                    teleport: true,
                },
            },
        })
    })

    afterEach(() => {
        wrapper.unmount()
    })

    describe('Integração/Componente', () => {
        describe('👀 Renderização:', () => {
            test('Dado o componente Quando renderizado Então deve possuir html equivalente ao Snapshot gravado', async () => {
                expect(wrapper.html()).toMatchSnapshot()
            })

            test('Dado o componente quando for montado deve ser uma instancia vue', () => {
                expect(wrapper.vm).toBeDefined()
            })
        })
        describe('🧠 Comportamento:', () => {
            test('Dado o botão de deslogar da conta Quando não houver nome do usuário Então mostra três pontos', async () => {
                mockStore.state.isLogged = true
                await nextTick()
                const buttonLogout = wrapper.get('#logout-button')

                expect(buttonLogout.text()).toBe('... (sair)')
            })

            test('Dado o botão de deslogar da conta Quando houver usuário Então deve exibir seu nome', async () => {
                mockStore.state.isLogged = true
                mockStore.state.userLogged.name = 'Jorge'
                await nextTick()
                const buttonLogout = wrapper.get('#logout-button')

                expect(buttonLogout.text()).toBe('Jorge (sair)')
            })

            test('Dado que o usuário precisa criar uma conta quando clicar em "Crie uma conta" então deve exibir um modal', async () => {
                mockStore.state.isLogged = false
                await nextTick()
                const button = wrapper.find('[data-create]')
                await button.trigger('click')
                const modalRegister = wrapper.get('[data-modal="create"]')

                expect(wrapper.emitted().click).toBeTruthy()
                expect(modalRegister.attributes('open')).toBe('true')
            })

            test('Dado que o usuário precisa efetuar login quando clicar em "Entrar" então deve exibir um modal', async () => {
                mockStore.state.isLogged = false
                await nextTick()
                const button = wrapper.find('[data-login]')
                await button.trigger('click')
                const modalLogin = wrapper.get('[data-modal="login"]')

                expect(wrapper.emitted().click).toBeTruthy()
                expect(modalLogin.attributes('open')).toBe('true')
            })
        })
        describe('🐕 Navegação:', () => {
            it.skip('deveria abrir a rota "Credentials" ao efetuar login', () => {
                return
            })
        })
    })
})
