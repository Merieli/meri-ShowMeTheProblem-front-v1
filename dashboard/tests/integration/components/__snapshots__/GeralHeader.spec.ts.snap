// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeralHeader IntegraÃ§Ã£o/Componente ðŸ‘€ RenderizaÃ§Ã£o: Dado o componente Quando renderizado EntÃ£o deve possuir html equivalente ao Snapshot gravado 1`] = `
"<div class=\\"bg-brand-main w-full justify-center flex\\">
  <header class=\\"geral-header justify-between items-center flex\\">
    <h1 class=\\"geral-header__logo font-black\\"><a href=\\"#\\">ShowMeTheProblem</a></h1>
    <nav class=\\"geral-header__navigation\\"><button id=\\"button-create\\" class=\\"geral-header__link font-bold\\" data-create=\\"\\"> Crie uma conta </button><button class=\\"base-button rounded-full font-bold text-xl transition-all duration-150 base-button_light\\" type=\\"button\\" data-button=\\"base\\" id=\\"button-login\\" data-login=\\"\\">
        <!-- @slot Texto de conteÃºdo do botÃ£o --> Entrar
      </button></nav>
  </header>
  <div data-modal=\\"create\\" type-modal=\\"create\\" open=\\"false\\">Modal de login e registro</div>
  <div data-modal=\\"login\\" type-modal=\\"login\\" open=\\"false\\">Modal de login e registro</div>
</div>"
`;
